src/dataflow/src/render/context.rs

```diff
@@ -333,40 +338,41 @@ where
+            use timely::scheduling::Activator;
+            let activations = trace.stream.scope().activations();
+            let activator = Activator::new(&info.address[..], activations);
+            // Maintain a list of work to do, cursor to navigate and process.
+            let mut todo = std::collections::VecDeque::new();
             move |input, output| {
+                // First, dequeue all batches.
                 input.for_each(|time, data| {
-                    let mut session = output.session(&time);
-                    for wrapper in data.iter() {
-                        let batch = &wrapper;
-                        let mut cursor = batch.cursor();
-                        if let Some(key) = &key {
-                            cursor.seek_key(batch, key);
-                            if cursor.get_key(batch) == Some(key) {
-                                while let Some(val) = cursor.get_val(batch) {
-                                    cursor.map_times(batch, |time, diff| {
-                                        for datum in logic(RefOrMut::Ref(val), time, diff) {
-                                            session.give(datum);
-                                        }
-                                    });
-                                    cursor.step_val(batch);
-                                }
-                            }
-                        } else {
-                            while let Some(_key) = cursor.get_key(batch) {
-                                while let Some(val) = cursor.get_val(batch) {
-                                    cursor.map_times(batch, |time, diff| {
-                                        for datum in logic(RefOrMut::Ref(val), time, diff) {
-                                            session.give(datum);
-                                        }
-                                    });
-                                    cursor.step_val(batch);
-                                }
-                                cursor.step_key(batch);
-                            }
-                        }
+                    let capability = time.retain();
+                    for batch in data.iter() {
+                        // enqueue a capability, cursor, and batch.
+                        todo.push_back(PendingWork::new(
+                            capability.clone(),
+                            batch.cursor(),
+                            batch.clone(),
+                        ));
                     }
                 });
+
+                // Second, make progress on `todo`.
+                let mut fuel = refuel;
+                while !todo.is_empty() && fuel > 0 {
+                    todo.front_mut()
+                        .unwrap()
+                        .do_work(&key, &mut logic, &mut fuel, output);
+                    if fuel > 0 {
+                        todo.pop_front();
```

Comment:
I'd consider asserting here that the `PendingWork` removed is indeed completed. It looks like it is from the `do_work` implementation, but it's always good to avoid surprises.

Reply:
Like, some sort of assert that if you try and read from or step the cursor you get nothing back?